Modules/_testcapi/getargs.c

```diff
@@ -333,68 +333,83 @@ getargs_K(PyObject *self, PyObject *args)
 static PyObject *
 test_k_code(PyObject *self, PyObject *Py_UNUSED(ignored))
 {
-    PyObject *tuple, *num;
-    unsigned long value;
-
-    tuple = PyTuple_New(1);
+    PyObject *tuple = PyTuple_New(1);
     if (tuple == NULL) {
         return NULL;
     }
 
     /* a number larger than ULONG_MAX even on 64-bit platforms */
-    num = PyLong_FromString("FFFFFFFFFFFFFFFFFFFFFFFF", NULL, 16);
+    PyObject *num = PyLong_FromString("FFFFFFFFFFFFFFFFFFFFFFFF", NULL, 16);
     if (num == NULL) {
-        return NULL;
+        goto error;
     }
 
-    value = PyLong_AsUnsignedLongMask(num);
-    if (value != ULONG_MAX) {
+    unsigned long value = PyLong_AsUnsignedLongMask(num);
+    if (value == (unsigned long)-1 && PyErr_Occurred()) {
+        Py_DECREF(num);
+        goto error;
+    }
+    else if (value != ULONG_MAX) {
+        Py_DECREF(num);
         PyErr_SetString(PyExc_AssertionError,
             "test_k_code: "
             "PyLong_AsUnsignedLongMask() returned wrong value for long 0xFFF...FFF");
-        return NULL;
+        goto error;
     }
 
     PyTuple_SET_ITEM(tuple, 0, num);
 
     value = 0;
     if (!PyArg_ParseTuple(tuple, "k:test_k_code", &value)) {
-        return NULL;
+        goto error;
     }
     if (value != ULONG_MAX) {
         PyErr_SetString(PyExc_AssertionError,
             "test_k_code: k code returned wrong value for long 0xFFF...FFF");
-        return NULL;
+        goto error;
     }
 
-    Py_DECREF(num);
+    Py_DECREF(tuple);  // also clears `num`
+    tuple = PyTuple_New(1);
+    if (tuple == NULL) {
+        return NULL;
+    }
```

Comment:
I don't like that we are repeating the same test with C&P code and slightly different parameters. IMO, it would be better if we factored out a helper function, and simply called that using a different set of params twice. What do you think?

Reply:
They are quite different for me:
<img width="1464" alt="Снимок экрана 2023-10-10 в 16 22 14" src="https://github.com/python/cpython/assets/4660275/2b7914c7-32ec-4599-84d3-5535252fde78">

There are some structural similarities, but not exactly 1-to-1.
<img width="1464" alt="Снимок экрана 2023-10-10 в 16 23 08" src="https://github.com/python/cpython/assets/4660275/2aeb5a47-22af-408c-b2f2-5b2803a17b7a">

I think that creating abstractions for just two functions would make C code harder, not easier. In Python code I would agree to abstract away common parts.